### DISCUSSION TITLE
Why not prefer $context over the SerializerAwareInterface to avoid a few design issues with the serializer?

### DISCUSSION THREAD

**Discussion by pieterw2w** [symfony/symfony#45541] [Q&A]
I think the symfony serializer design can be improved a little bit how recursion works. Right now recursive serialization is done with use of the SerializerAwareInterface and the SerializerAwareTrait. There are quite a few problems with this setup when writing my own normalizer which were all related to the use of the SerializerAwareInterface. First of all you can not reuse the same class instance for multiple serializers because it will internally only use the last serializer it was linked to:

```php
$serializerAwareNormalizer = new GetSetMethodNormalizer();
$firstSerializer = new Serializer([$serializerAwareNormalizer]);
$secondSerializer = new Serializer([$serializerAwareNormalizer]);
```

Now if I would call $firstSerializer->normalize(..) I would magically switch to the $secondSerializer if recursion is involved.

Another issue is that I want to map errors to a validation error and tell where the error is, but the problem is that a normalizer does not know what the parent object that was serialized let alone the parent keys that were followed, so I can not construct a good error message.
I could work around by making my own object normalizer, implement SerializerAwareInterface and add things to the $context variable, but the problem here is that this means all the normalizers require to update this context variable.

Ideally I would expect this syntax over the SerializerAwareInterface to handle recursion properly(this is just an example class with usecases how a SerializerContext object could work):

```php
class RecursiveNormalizerExample implements NormalizerInterface
{
    public function normalize($data, SerializerContext $context)
    {
        $format = $context->getFormat(); // currently this is an argument in the NormalizerInterface that is rarely used in normalizers.
        $maxDepth = $context->getValue(AbstractObjectNormalizer::ENABLE_MAX_DEPTH);
        if (empty($data->child1)) {
            throw new RuntimeException('Error at ' . implode(',', $context->getRecursionKey()));
        }
        return [
            'format' => $format,
            'maxDepth' => $maxDepth,
            'value1' => $context->normalizeChild('value1', $data->child1),
            'value2' => $context->normalizeChild('value2', $data->child2),
        ];
    }
}
```
Of course writing it like this would be a huge BC break, so maybe the BC solution would be $context['serializer'] or something like that. But I think this will improve the serializer component a lot and also useful for the API Platform.

Nobody likes to debug serialization of big (recursive) objects with exception messages that can not tell where the serialization problem is.

**Comment by lyrixx** [comment]
> First of all you can not reuse the same class instance for multiple serializers because it will internally only use the last serializer it was linked to:

you are right. The main idea behind the serializer is to use only one serializer, and attach many normalizer to it.

About the validation, you have already existing in place in symfony to do that. For example: https://speakerdeck.com/lyrixx/symfony-live-2017-serializer?slide=62

An from SF 5.4, you also get error about typehint: https://symfony.com/blog/new-in-symfony-5-4-serializer-improvements#collect-denormalization-type-errors

Finally, I usually use the decorator pattern to handle the normalization of a specific type. This avoid me to use `SerializerAwareInterface` hack.

---

I understand I not really answered your questions, but I hope it may help you anyway.

**Reply by pieterw2w** [reply]
I'm aware I could use the symfony validator, but the problem lies in that the validator only works on existing objects. If you are into avoiding anemic domain objects, then you can't have invalid objects, so the only way to make validator work like this is having a DTO just for the validation part that should match the actual domain object, which sounds like a very fragile construction.

In a nutshell, imagine I have something like this:
```php
<?php
class AnotherSubObject {
    private int $id;
    
    public function __construct(int $id)
    {
        $this->id = $id;
    }
}

class SubObject {
    public AnotherSubObject $key;
}

class Example {
    public SubObject $subObject;
    public ?SubObject $anotherSubObject = null;
}
```

If I try to normalize(['subObject' => ['key' => []]) it will result in a 500 error that 'id' is required, but it's not possible to respond with a validation error that 'subObject.key.id' is required, which would make it much easier to debug.

If I try to normalize(['subObject' => ['key' => ['id' => 'string']]) it will result in a 500 error that 'string' can not be serialized to an integer, but it's not possible to respond with a validation error that 'subObject.key.id' is the wrong type.

From an API point of view these should be validation errors and not 500 errors. Plus if your object is quite big with some recursion it is no longer easy to figure out where the error is.

I ended up writing my own object normalizer (and AbstractObjectNormalizer could not be used in my case either) and adding the recursion information in $context, but I still ran into the issue that the serializer does not keep track of it's recursion itself and other (vendor) normalizers not updating my custom recursion information in $context obviously.
